### PR TITLE
X img diff tool

### PR DESCRIPTION
**Pull request by langolf**
Let's try again!

### CODE REVIEW

**Comment by Winner95** [comment]
@langolf  thank you! I will look into some details today and will get with results asap)

fyi @NikhilVerma

File: src/bin.ts [hunk 1]
```diff
@@ -3,50 +3,14 @@
 import { argv } from 'yargs';
 import path from 'path';
 import runVrt from './index';
```
Comment (Winner95) [inline src/bin.ts:5]: import runVrt from './index';
import { VRTCommandOptions } from './index';

can be:

import runVrt, { VRTCommandOptions } from './index';

also `madge --circular src/bin.ts`

Found 1 circular dependency:

1) index.ts > report.ts

could you please update it?

File: src/report.ts [hunk 2]
```diff
@@ -1,11 +1,11 @@
 import fs from 'fs';
-import type { DiffResult } from 'compare';
+import { VRTResultsEntries } from 'index';
```
Comment (Winner95) [inline src/report.ts:2]: this is circular dependency. You can export VRTResultsEntries to separate file and import it, where it is needed.

File: src/bin.ts [hunk 3]
```diff
@@ -3,50 +3,14 @@
...
-const options = {
-    threshold,
-    includeAA,
-    alpha,
-    aaColor,
-    diffColor,
-    diffColorAlt,
-    diffMask,
-};
+const options = (argv.options as VRTCommandOptions) || {};
```
Comment (Winner95) [inline src/bin.ts:10]: currently if we try to pass --options and any value - it will return boolean `true` value
we probably need to have opportunity to reset each option value - please check implementation before - I had exact the same issue  before)

**Review by Winner95 — CHANGES_REQUESTED** [review]

File: src/bin.ts (hunk 3 above)
Comment (Winner95) [reply src/bin.ts:10]: let's add it in the readme

**Review by Winner95 — CHANGES_REQUESTED** [review]

**Comment by Winner95** [comment]
@langolf - teamcity arg is most crucial note. The fallback for threshold is also important. the rest is codestyle.

fyi: @NikhilVerma

File: src/bin.ts [hunk 4]
```diff
@@ -1,52 +1,32 @@
...
+        'strip-dashed': true,
+        'camel-case-expansion': false,
+    })
+    .strict().argv;
 
-    if (rgb.length !== 3) {
-        console.log(
-            `warning: color ${color} is not matching scheme of [number, number, number]. It won't be used in options`
-        );
+const regCliOptions: [string, any][] = Object.entries(options).slice(1, -1); // yargs extra keys
```
[the remark this answers is not in the document]
Comment (langolf) [reply src/bin.ts:30]: Unfortunately no. Yargs validating options not in runtime. It is just tech info for us.

File: README.md [hunk 5]
```diff
@@ -36,8 +46,6 @@ import runVrt from '@magiclab/vrt-runner';
 runVrt({
     cwd,
     output,
```
Comment (Winner95) [inline README.md:48]: please return teamcity and options to the readme as well.

When you edit files - please refer to diff change in github diff online - it help you to see produced diff and remove unnecessary changes.

File: README.md [hunk 6]
```diff
@@ -48,8 +56,7 @@ Currently we support `onVrtComplete` hook, which allows you to get results of co
 One of the ways to use this data is the following:
 
 ```js
-
-(async function() {
+(async function () {
```
Comment (Winner95) [inline README.md:59]: if you use readme formatter - could you provide a name of it, because it might be that we use different one and we will keep producing diff in this file.

The idea is to add it to the precommit hook or github action to this repo. It should be a separate pull-request. 

Meanwhile, could you switch this off? 

These formmating changes are not necessary.

File: package.json [hunk 7]
```diff
@@ -37,15 +35,16 @@
     "build": "rm -rf build && mkdir -p build && cp -R src/report-assets build && ./node_modules/.bin/tsc",
     "styles": "npx node-sass --no-source-map ./src/report-assets/css/vrt.scss ./src/report-assets/css/vrt.css --output-style compressed",
     "tsc:lint": "tsc --incremental --noEmit false",
-    "test": "npm run build && node build/bin.js --cwd ./test/1/ --output ./tmp/result/",
-    "test:debug": "npm run build && node --inspect --inspect-brk build/bin.js --cwd ./test/1/ --output ./tmp/result/",
+    "test:clean": "rm -rf tmp/result",
+    "test": "npm run build && npm run test:clean && node build/bin.js --cwd test/1 --output tmp/result",
+    "test:debug": "npm run build && node --inspect --inspect-brk build/bin.js --cwd test/1/ --output tmp/result/",
```
[the remark this answers is not in the document]
Comment (Winner95) [reply package.json:40]: @langolf  we probably need to clean folder in debug as well?

File: src/bin.ts [hunk 8]
```diff
@@ -1,52 +1,33 @@
 #!/usr/bin/env node
 
-import { argv } from 'yargs';
```
Comment (Winner95) [inline src/bin.ts:3]: I just spotted smth - why we renamed half of the variables or moved their positions? it makes no difference, apart from diff size. 

Could you please keep yargs on 3rd line, please?

File: src/bin.ts [hunk 9]
```diff
@@ -1,52 +1,33 @@
...
+import yargs from 'yargs';
 import runVrt from './index';
 
-function createRgbArrayFromString(color: string): undefined | [number, number, number] {
-    const rgb = color
-        .replace(/[[\]]/g, '')
-        .split(',')
-        .map((item) => parseInt(item))
-        .filter((item) => !isNaN(item));
+const { cwd, output, teamcity, ...rest } = yargs(process.argv.slice(2))
```
Comment (Winner95) [inline src/bin.ts:7]: why `rest` - not options? mentioned before - this renaming is not necessary. (2)

File: src/bin.ts [hunk 10]
```diff
@@ -1,52 +1,33 @@
...
+    .options({
+        cwd: { type: 'string', default: process.cwd(), demandCommand: true },
+        teamcity: { type: 'boolean', default: true },
```
Comment (Winner95) [inline src/bin.ts:10]: it wasn't `true` by default - let's keep changes to the minimum.

File: src/index.ts [hunk 11]
```diff
@@ -1,61 +1,37 @@
 import path from 'path';
-import diffDirs, { DirsType, DiffResult, ComparisonOptionsType } from './compare';
-import generateReport from './report';
+import generateReportTemplate from './report';
```
Comment (Winner95) [inline src/index.ts:2]: unnecessary renaming `generateReport`

File: src/index.ts [hunk 12]
```diff
@@ -1,61 +1,37 @@
...
-    const info = showResults({
-        failed: result.failed.length + result.missing.length,
-        passed: result.passed.length + result.new.length,
-        diffTime: cmpTime / 1000,
-    });
-
-    console.info(info);
-};
+// Special log messages for TeamCity service. Using by other platforms
+function teamcityLogger(message: string) {
```
Comment (Winner95) [inline src/index.ts:8]: can we reuse `teamcityMessage` function from compare - move it out from lexical scope of `diffDirs` to the root level and export it to this file?

File: src/index.ts [hunk 13]
```diff
@@ -1,61 +1,37 @@
...
-    teamcity,
     output,
     cwd,
-    onVrtComplete,
     options,
 }: {
-    teamcity: boolean;
     output: string;
     cwd: string;
-    onVrtComplete?: onVrtCompleteType;
```
Comment (Winner95) [inline src/index.ts:48]: why did you delete this hook? It is for node.js only, but we need to work with data outside of the script scope, please revert this change.

File: src/index.ts [hunk 14]
```diff
@@ -1,61 +1,37 @@
...
-    const reportFile = path.resolve(output, 'index.html');
-
-    const dirs: DirsType = {
+    const dirs = {
         baseline: path.resolve(output, 'baseline'),
         test: path.resolve(output, 'test'),
         diff: path.resolve(output, 'diff'),
     };
 
+    const reportPage = path.resolve(output, 'index.html');
```
Comment (Winner95) [inline src/index.ts:28]: why we changed position of this file here? please revert this change

File: src/index.ts [hunk 15]
```diff
@@ -1,61 +1,37 @@
...
     cwd: string;
-    onVrtComplete?: onVrtCompleteType;
-    options?: ComparisonOptionsType;
+    teamcity: boolean;
+    options: [string, any][];
 }) {
-    const reportFile = path.resolve(output, 'index.html');
-
-    const dirs: DirsType = {
+    const dirs = {
```
Comment (Winner95) [inline src/index.ts:22]: does type now resolved correctly without assigning?

File: src/index.ts [hunk 16]
```diff
@@ -69,26 +45,51 @@ export default async function runVrt({
         fs.copySync(path.resolve(cwd, 'test'), dirs.test);
     }
 
-    // Collect images from baseline and testing paths
```
Comment (Winner95) [inline src/index.ts:72]: do we need this change for this PR?

File: src/index.ts [hunk 17]
```diff
@@ -69,26 +45,51 @@ export default async function runVrt({
...
+        teamcityLogger(`testSuiteStarted name='VRT'`);
         let cmpTime = Date.now();
-        const result = await diffDirs({
-            dirs,
-            teamcity,
-            options,
-        });
+
+        // we have to convert to strings
+        // const regFlags = options.map((option) => option.map((item) => `--${item[0]}=${item[1]}`));
```
Comment (Winner95) [inline src/index.ts:53]: extra comment?

File: src/index.ts (hunk 14 above)
Comment (Winner95) [reply src/index.ts:28]: please remove renaming as well.

File: src/index.ts [hunk 18]
```diff
@@ -69,26 +45,51 @@ export default async function runVrt({
...
+            failedItems: string[];
+            passedItems: string[];
+            newItems: string[];
+            deletedItems: string[];
+        } = JSON.parse(fs.readFileSync(vrtCommandReportFile, 'utf8'));
+
+        if (vrtReportStats.failedItems.length > 0) {
+            for (const file of vrtReportStats.failedItems) {
+                teamcityLogger(
+                    `testFailed name='${file}' message='sandbox screenshots are different' details=''`
```
Comment (Winner95) [inline src/index.ts:78]: can we provide details - threshold at which diff failed?

File: src/report.ts [hunk 19]
```diff
@@ -1,11 +1,11 @@
...
+    newItems: string[];
+    deletedItems: string[];
+};
 
-export default function generateReport(reportFile: string, vrtResult: DiffResult) {
-    fs.writeFileSync(reportFile, html(vrtResult));
-}
-
-function html(vrtResult: DiffResult) {
+export default function generateReportTemplate(data: VRTResultsEntries) {
```
Comment (Winner95) [inline src/report.ts:8]: do we need this change in scope of this PR?  please revert it

File: src/report.ts [hunk 20]
```diff
@@ -19,7 +19,7 @@ function html(vrtResult: DiffResult) {
         <body class="page-comparison">
             <div class="vrt-wrapper">
                 ${navigation()}
-                ${main(vrtResult)}
+                ${main(data)}
```
Comment (Winner95) [inline src/report.ts:22]: It looks like we just renamed some variables, which are used once within this function scope. Please revert these changes.

Let's keep scope of this PR only to new diff-tool + new options, coming with it in bin + index.

**Review by Winner95 — CHANGES_REQUESTED** [review]

**Comment by Winner95** [comment]
hey @langolf , I looked via compared diff and discovered that after last changes, we have around 50% of changes are not related to diff tool and can be avoided for now. We also lost newly introduced hook `onVrtComplete`. Please, fix all the notes provided. 

I suggest that all structural and naming  and position of variables changes should be done separately or discussed in separate PR. It saves on your time and on review time.

Please let me know if you need any help with it:)

fyi: @NikhilVerma

**Comment by Winner95** [comment]
hey @NikhilVerma , could you please help us with merging these changes - we can now test changes from @langolf  in a new release.

@langolf  thank you for your contribution!
